Add specification draft

ulan opened the pull request:
Preview is available at https://ulan.github.io/misc/measure-memory-spec/

It can be generated locally using `bikeshed --die-on=warning watch index.src.html index.html`

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+Repository: WICG/performance-measure-memory
+Shortname: performance-measure-memory
+Markup Shorthands: markdown yes
+Markup Shorthands: css no
```
FYI you can combine these into `markdown yes, css no`

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+  text: JavaScript realm; type: interface; url: sec-code-realms
+  text: JavaScript realms; type: interface; url: sec-code-realms
+  text: JavaScript agent clusters; type: interface; url: sec-agent-clusters
+urlPrefix: https://www.w3.org/TR/hr-time-2; spec: HR-TIME-2
```
You should always reference editors' drafts, so this should be https://w3c.github.io/hr-time/

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+urlPrefix: https://www.w3.org/TR/hr-time-2; spec: HR-TIME-2
+  type: interface
+    text: Performance; url: #sec-performance
+  type: grammar
```
I don't think these are part of HR time. Probably remove them?

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+<div nonnormative>
+The tradeoff between memory and performance is inherent in many algorithms and data-structures.
+Web developers today have multiple ways to measure the timing information and no standard way to measure the memory usage.
+This specification defines a [=performance.measureMemory=] API that estimates the memory usage of the web application including all its iframes and workers.
```
It'd be best to link to the IDL definition, using `{{Performance/measureMemory()|performance.measureMemory()}}`.

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+
+
+<div class="example" id="example4">
+To get the memory usage of a shared/service worker, the `performance.measureMemory()`
```
Same cross-linking syntax here

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+
+<div class="example" id="example4">
+To get the memory usage of a shared/service worker, the `performance.measureMemory()`
+function needs to be invoked in the context of that worker.
```
Maybe add "The result could be something like:"

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+Examples {#examples}
+--------------------
+
+<div class="example" id="example1">
```
I suggest giving these meaningful IDs, so that if you reorder or remove them, they don't mismatch their labels. E.g. "example-basic-usage", "example-simple-page", "example-same-origin-iframe", ...

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+<pre class="idl">
+dictionary MemoryMeasurement {
+  required unsigned long long bytes;
+  required sequence&lt;MemoryBreakdownEntry&gt; breakdown;
```
Since these are output-only dictionaries, they should not use `required`: see https://heycam.github.io/webidl/#required-dictionary-member

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+};
+</pre>
+<dl class="domintro">
+:  <code>self . {{MemoryMeasurement/bytes}}</code>
```
When we use `self` in these boxes, we're referring to the global property. So you probably want to use something more specific like `measurement` here, and `breakdown` below.

ulan replied:
Good to know! Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+Memory measurement result {#memory-measurement-result}
+------------------------------------------------------
+
+The `performance.measureMemory()` function returns a {{Promise}} that resolves to an instance of `MemoryMeasurement` dictionary:
```
In general, any usage of backticks should likely be a cross-link instead. So `performance.measureMemory()` should be linked as above, and `{{MemoryMeasurement}}` should be used.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+:: A number that represents the total memory usage.
+
+:  <code>self . {{MemoryMeasurement/breakdown}}</code>
+:: A list that partitions the total {{MemoryMeasurement/bytes}} and provides attribution and type information. </dl>
```
I would say "array" instead of "list" or "set", since this is web-developer-facing documentation and arrays are something they're familiar with.

ulan replied:
Done. I kept "set" in other more formal parts of the spec.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+<pre class="anchors">
+urlPrefix: https://tc39.es/ecma262/; spec: ECMAScript
+  text: SharedArrayBuffer; type: interface; url: sec-sharedarraybuffer-objects
+  text: JavaScript realm; type: interface; url: sec-code-realms
```
These three should be `type: dfn` instead of `type: interface`. And thus links to them should use `[=JavaScript realm=]` instead of `{{JavaScript realm}}`.

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+
+<pre class="idl">
+dictionary MemoryAttribution {
+  DOMString url;
```
It doesn't really matter for output-only dictionaries, but for consistency, it'd be good to use `USVString` here, since it contains a URL. See https://url.spec.whatwg.org/#url-apis-elsewhere

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+
+:  <code>self . {{MemoryAttributionContainer/src}}</code>
+:: The src attribute of the container element. If the container element is an {{HTMLObjectElement}},
+   then this field contains the value of the {{HTMLObjectElement/data}} attribute.</dl>
```
It'd be slightly nicer to say `If the container element is an <{object}> element, then this field contains the value of the <{object/data}> attribute`. That's a bit more geared toward the developer-facing HTML concepts instead of the JavaScript object interface.

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+Top-level algorithms {#top-level-algorithms}
+--------------------------------------------
+<div algorithm>
+  <dfn>performance.measureMemory</dfn> must run the following steps:
```
This should be done as

```html
The <dfn method for="Performance">measureMemory()</dfn> method steps are:
```

Then you will get proper cross-linking from the IDL block, and any other references to `{{Performance/measureMemory()}}`, to this algorithm.

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+--------------------------------------------
+<div algorithm>
+  <dfn>performance.measureMemory</dfn> must run the following steps:
+  1. Let |realm| be the [=current settings object=]'s [=environment settings object/Realm=].
```
This should just be the `[=current Realm=]`, although I think that'll require another entry into your anchors block.

ulan replied:
Done, added an anchor.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+<div algorithm>
+  <dfn>performance.measureMemory</dfn> must run the following steps:
+  1. Let |realm| be the [=current settings object=]'s [=environment settings object/Realm=].
+  1. [=Assert=]: |realm|'s [=agent=]'s [=agent cluster=]'s [=cross-origin-isolated=] is true.
```
These aren't cross-linking in the output. You may want to use `--die-on=warning`.

To fix it, I think you could add aliases for "JavaScript agent" and "JavaScript agent cluster" to your anchors block?

For cross-origin isolated, it's actually associated with the browsing context group. So I think you need realm's agent's agent cluster's browsing context group's cross-origin isolated. (Also, only one hyphen.)

Even that won't work, I think, because HTML does not yet export "cross-origin isolated". So you could add it to your anchors block. Eventually we should export it though so you don't need to do that...

ulan replied:
Added anchors for [=agent=], [=agent cluster=], [=cross-origin isolated=].

I have to use [the cross-origin isolated flag](https://html.spec.whatwg.org/multipage/webappapis.html#cross-origin-isolated) of the agent cluster because the method is also defined for shared/service workers. Is WAI that the flag is defined both for a browsing context group and an agent cluster?

ulan replied:
I changed the anchors a bit to make it clear that they refer to fields: e.g. `[agent cluster/cross-origin isolated]`.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+  1. [=Assert=]: |realm|'s [=agent=]'s [=agent cluster=]'s [=cross-origin-isolated=] is true.
+  1. Let |global object| be the [=current global object=].
+  1. Let |promise| be a new {{Promise}}.
+  1. If [=memory measurement allowed predicate=] given |global object| and [=current settings object=] is false, then:
```
It's weird that this algorithm takes both a global object and an environment settings object, when they're 1:1.

Also this algorithm constantly switching back and forth between realms/ESOs/global objects feels a bit excessive. Ultimately you do need to use all three in this spec (which is kind of unique), but maybe in this top-level algorithm you could reduce things a bit, I'm not sure...

ulan replied:
Refactored the algorithm to use realms consistently instead of switch to ESO and global objects.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+  1. Let |global object| be the [=current global object=].
+  1. Let |promise| be a new {{Promise}}.
+  1. If [=memory measurement allowed predicate=] given |global object| and [=current settings object=] is false, then:
+      1. Reject |promise| with a "Security Error" {{DOMException}}.
```
```suggestion
      1. Reject |promise| with a "{{SecurityError}}" {{DOMException}}.
```

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+  1. Let |global object| be the [=current global object=].
+  1. Let |promise| be a new {{Promise}}.
+  1. If [=memory measurement allowed predicate=] given |global object| and [=current settings object=] is false, then:
+      1. Reject |promise| with a "Security Error" {{DOMException}}.
```
Probably you want to return |promise|, instead of continuing? Alternately this could return `[=a promise rejected with=] a...` and delay creating the promise until right before you use it.

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+      1. Reject |promise| with a "Security Error" {{DOMException}}.
+  1. Let |agent clusters| be the result of [=getting all agent clusters=] given |realm|.
+  1. Start asynchronous [=implementation-defined memory measurement=] given |agent clusters| and |promise|.
+  1. Return |promise|
```
```suggestion
  1. Return |promise|.
```

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+<div algorithm>
+  To <dfn lt="getting all agent clusters|get all agent clusters">get all agent clusters</dfn>
+  given an {{JavaScript realm}}  |realm|:
+  1. If |realm|'s [=global object=] is a {{Window}}, then:
```
This links to the general "global object" concept, but you want to link to `[=realm/global object=]`.

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+  To <dfn lt="getting all agent clusters|get all agent clusters">get all agent clusters</dfn>
+  given an {{JavaScript realm}}  |realm|:
+  1. If |realm|'s [=global object=] is a {{Window}}, then:
+      1. Let |browsing context| be |realm|'s [=settings object=]'s [=responsible document=]'s [=Document/browsing context=].
+      1. Let |agent clusters| be the result of [=map/getting the values=] of |browsing context|'s
+         [=browsing context group=]'s [=agent cluster map=].
+      1. Return |agent clusters|.
```
You can get from a Window to its browsing context more easily via `[=Window/browsing context=]`.

There's a slight problem here that in specs we don't have a rigorous pointer from browsing context to group, except for top-level browsing contexts. But we can probably hand-wave that a bit; it's obvious what is meant. I'd say

```html
1. Let |group| be the [=browsing context group=] that contains |realm|'s [=realm/global object=]'s [=Window/browsing context=].
1. Return the result of [=map/getting the values=] of |group|'s [=agent cluster map=].
```

ulan replied:
Done, thanks!

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+      1. Let |agent clusters| be the result of [=map/getting the values=] of |browsing context|'s
+         [=browsing context group=]'s [=agent cluster map=].
+      1. Return |agent clusters|.
+  1. Return a singleton [=set=] of |realm|'s [=agent=]'s [=agent cluster=].
```
I'd say "Return  « realm's agent's agent cluster »."

ulan replied:
Done. Good to know that «» refers to a literal.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+
+<div algorithm>
+  To perform <dfn>implementation-defined memory measurement</dfn> given a [=set=] of [=agent clusters=] |agent clusters| and
+  a {{Promise}} |promise| run these steps asynchronously:
```
I think "asynchronously" should be `[=in parallel=]`?

ulan replied:
I agree that [=in parellel=] fits better here. I move the step that creates the measurement result to the task that resolves the promise so that observable object creation happens on the main thread.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+<div algorithm>
+  To perform <dfn>implementation-defined memory measurement</dfn> given a [=set=] of [=agent clusters=] |agent clusters| and
+  a {{Promise}} |promise| run these steps asynchronously:
+  1. Let |intermediate memory measurement| be implementation-defined memory measurement of |agent clusters|.
```
This should probably be something like "... be an [=implementation-defined=] [=intermediate memory measurement=] of..."

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+  a {{Promise}} |promise| run these steps asynchronously:
+  1. Let |intermediate memory measurement| be implementation-defined memory measurement of |agent clusters|.
+  1. Let |memory measurement| be the result of [=creating a new memory measurement=] given |intermediate memory measurement|.
+  1. Queue a task that resolves |promise| with |memory measurement|.
```
Task-queuing needs a bit more specificity. I think you want `[=Queue a global task=] on the ??? task source given |promise|'s [=relevant global object=] to [=resolve=] |promise| with |memory measurement|.`

As for picking a task source... what task source does Chromium use?

ulan replied:
In Chromium we use V8's task queue. (The same queue runs the cleanup tasks of JS WeakRefs). Is there a default task queue in the spec that I could use? Relation and order w.r.t to other tasks is not important for the API.

I left a TODO here.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+  1. Let |token| be |realm|'s [=global object=]'s [=WindowOrWorkerGlobalScope/memory attribution token=].
+  1. If |token|'s [=memory attribution token/cross-origin aggregated flag=] is true, then
+      1. Return a new {{MemoryAttribution}} whose:
+          - {{MemoryAttribution/url}} is "cross-origin-url",
```
Very much a nit:

```suggestion
          - {{MemoryAttribution/url}} is "`cross-origin-url`",
```

and similar for other strings.

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+  1. If |token|'s [=memory attribution token/cross-origin aggregated flag=] is true, then
+      1. Return a new {{MemoryAttribution}} whose:
+          - {{MemoryAttribution/url}} is "cross-origin-url",
+          - {{MemoryAttribution/container}} is |token|'s [=memory attribution token/container=] if it is not null,
```
What if it is null? Is the intention to omit the dictionary entry? It might be good to state that explicitly.

FYI you could also do this using Infra ordered map manipulation (E.g. `Let |returnValue| be «[ "{{MemoryAttribution/url}}" → "`cross-origin-url`", ... ]». If X, set |returnValue|["{{MemoryAttribution/container}}"] to...`.) But I think your way is nice and easy to read, and just needs a minor clarification on this point.

ulan replied:
Done. Added "otherwise, the {{MemoryAttribution/container}} entry is omitted".

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+</div>
+
+<div algorithm>
+  To <dfn lt="get the global scope name|getting the global scope name">get the global scope name</dfn>
```
I think this algorithm can be eliminated and replaced with `The [=identifier=] of |scope|'s [=interface=]`. See the precedent in https://heycam.github.io/webidl/#create-an-interface-object step 5.

ulan replied:
Nice, thanks! Replaced all with
`Let |scope name| be [=identifier=] of |realm|'s [=Realm/global object=]'s [=interface=].`

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+<div algorithm>
+  To <dfn lt="extract container element attributes|extracting container element attributes">extract container element attributes</dfn>
+  given an {{HTMLElement}} |container element|:
+  1. Switch on |container element|'s {{Element/tagName}}:
```
You should use `[=local name=]` instead of the public API `tagName`.

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+      <dl class="switch">
+        : "iframe"
+        :: Return a new {{MemoryAttributionContainer}} whose:
+            - {{MemoryAttributionContainer/id}} is |container element|'s {{Element/id}},
```
Instead of using the public JS APIs in this way, use the backing concepts. In this case it's the HTML attribute values, so `... is the value of |container element|'s <{???/id}>` attribute`.

For the ???, see https://github.com/whatwg/html/issues/5417. Maybe I can fix that soon. In the meantime you could patch it up with the anchor block, or just use backticks instead of cross-linking for now.

ulan replied:
I created anchors for element/id and frame/src, other attributes have proper links.
I had to use `type:dfn` since I couldn't figure out how to make `type:element-sub` to work.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+};
+</pre>
+<dl class="domintro">
+:  <code>self . {{Performance/measureMemory}}</code>
```
```suggestion
:  <code>self . {{Performance/measureMemory}}()</code>
```

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+Integration with the existing specification {#integration-with-the-existing-specification}
+==========================================================================================
+
+Extensions to the {{Performance}} interface {#extensions-to-performance}
```
Since this definition is the main content of this spec, I'd say it doesn't need to be in the "Integrations" section. Instead you can move it up to the processing model section, right above the actual definition of the `performance.measureMemory()` algorithm itself.

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+</pre>
+<dl class="domintro">
+:  <code>self . {{Performance/measureMemory}}</code>
+:: A method that performs an asynchronous memory measurement. </dl>
```
Maybe cross-reference folks to `[[#memory-measurement-result]]` if they want more details on what it returns?

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+====================================
+Thanks to Domenic Denicola and Shu-yu Guo for contributing to the API design and for reviewing this specification.
+
+Without implying any endorsement of this specification, thanks to Anne van Kesteren, Boris Zbarsky, Dominik Inführ, Chris Hamilton, Hannes Payer, Joe Mason, Kentaro Hara, L. David Baron, Mathias Bynens, Matthew Bolohan, Michael Lippautz, Neil Mckay Olga Belomestnykh, Per Parker, Philipp Weis, and Yoav Weiss, for their feedback and contributions.
```
I think omitting the disclaimer is OK :). It's generally understood that acknowledgments don't imply endorsement.

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,797 @@
+----------------------------------------------------------------------------------------------------
+
+<div algorithm>
+  To <dfn lt="obtain window memory attribution token|obtaining window memory attribution token">obtain window memory attribution token</dfn>
```
Nit: I'd put an "a" in the name, i.e. "obtain a window memory attribution token"

ulan replied:
Done.

ulan commented:
Thanks a lot for reviewing, @domenic! Happy to hear that the draft is in a good shape. This made my day.

I may be a bit slow with addressing the comments until Wednesday.

ulan commented:
Thanks the comments, @domenic. I learned a lot from them and bikeshed now runs without warnings.

Please take another look. I uploaded the compiled html version [here](https://ulan.github.io/misc/measure-memory-spec/).

domenic commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+      bytes: 500000,
+      attribution: [
+        {
+          url: 'https://example.com/iframe.html'
```
Nit: you started using a mix of single and double quotes here

ulan replied:
Done

domenic commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+:: An array of [=JavaScript realms=] specified by their URLs and/or container elements that use the memory.
+
+:  <code>breakdown . {{MemoryBreakdownEntry/userAgentSpecificTypes}}</code>
+:: An array of implementation defined memory types associated with the memory. </dl>
```
```suggestion
:: An array of [=implementation-defined=] memory types associated with the memory. </dl>
```

domenic commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+:: A [=set=] of [=JavaScript realms=] to which the memory is attributed to.
+
+:  <dfn for="intermediate memory breakdown entry">user agent specific types</dfn>
+:: A [=set=] of [=/strings=] specifying implementation defined memory types associated with the memory.
```
```suggestion
:: A [=set=] of [=/strings=] specifying [=implementation-defined=] memory types associated with the memory.
```

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+Intermediate memory measurement {#intermediate-memory-measurement-section}
+--------------------------------------------------------------------------
+
+This specification assumes the existences of an implementation-defined algorithm that can measure
```
```suggestion
This specification assumes the existences of an [=implementation-defined=] algorithm that can measure
```

domenic commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+
+:  <code>attribution . {{MemoryAttribution/container}}</code>
+:: Describes the DOM element that (maybe indirectly) contains the [=JavaScript realms=].
+   It may be empty if the attribution is for the same-origin top-level realm.
```
"It may be empty" is a bit confusing. I think this should be "This property might be absent"?

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+:  <code>attribution . {{MemoryAttribution/scope}}</code>
+:: Describes the type of the same-origin [=JavaScript realm=]:
+   <code>"Window", "DedicatedWorkerGlobalScope", "SharedWorkerGlobalScope", "ServiceWorkerGlobalScope"</code>
+   or contains <code>"cross-origin-aggregated"</code> for the cross-origin case. </dl>
```
```suggestion
   or <code>"cross-origin-aggregated"</code> for the cross-origin case. </dl>
```

domenic commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+</pre>
+<dl class="domintro">
+:  <code>container . {{MemoryAttributionContainer/id}}</code>
+:: The id attribute of the container element.
```
```suggestion
:: The `id` attribute of the container element.
```

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+:: The id attribute of the container element.
+
+:  <code>container . {{MemoryAttributionContainer/src}}</code>
+:: The src attribute of the container element. If the container element is an <{object}> element,
```
```suggestion
:: The `src` attribute of the container element. If the container element is an <{object}> element,
```

domenic commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+};
+</pre>
+<dl class="domintro">
+:  <code>self . {{Performance/measureMemory()}}</code>
```
```suggestion
:  <code>performance . {{Performance/measureMemory()}}</code>
```

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+--------------------------------------------
+<div algorithm>
+  The <dfn method for="Performance">measureMemory()</dfn> method steps are:
+  1. [=Assert=]: [=current Realm=]'s [=Realm/agent=]'s [=agent/agent cluster=]'s [=agent cluster/cross-origin isolated=] is true.
```
This assertion can fail currently, because we don't have https://github.com/heycam/webidl/pull/883 in use yet.

I might try to get that landed. But in the meantime, try adding an anchors block entry for the CrossOriginIsolated extended attribute, linking to the pull request (or maybe to https://pr-preview.s3.amazonaws.com/mikewest/webidl/pull/883.html#CrossOriginIsolated directly). Then update the IDL fragment to use it.

That's a little funky, but it seems like the right move.

ulan replied:
Good to know! Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+--------------------------------------------
+<div algorithm>
+  The <dfn method for="Performance">measureMemory()</dfn> method steps are:
+  1. [=Assert=]: [=current Realm=]'s [=Realm/agent=]'s [=agent/agent cluster=]'s [=agent cluster/cross-origin isolated=] is true.
```
Here and throughout, we usually say "the" current Realm.

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+  1. Let |scope name| be [=identifier=] of |realm|'s [=Realm/global object=]'s [=interface=].
+  1. Return a new {{MemoryAttribution}} whose:
+      - {{MemoryAttribution/url}} is |realm|'s [=Realm/settings object=]'s [=environment/creation URL=],
+      - {{MemoryAttribution/container}} is |token|'s [=memory attribution token/container=] if it is not null,
```
Port over the "otherwise..." from a couple lines up here too.

ulan replied:
Done.

domenic commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+<div algorithm>
+  To <dfn lt="obtain a window memory attribution token|obtaining a window memory attribution token">obtain a window memory attribution token</dfn>
+  given an [=/origin=] |origin|, an [=/origin=] |parent origin|, an [=/origin=] |top-level origin|,
+  an {{HTMLElement}} |container element|, an [=/memory attribution token=] |parent token|:
```
```suggestion
  an {{HTMLElement}} |container element|, and a [=/memory attribution token=] |parent token|:
```

domenic commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+    text: cross-origin isolation; url: multipage/webappapis.html#cross-origin-isolated
+    text: agent cluster map; for: browsing context group; url: multipage/browsers.html#agent-cluster-map
+    text: id; for: element; url: multipage/dom.html#the-id-attribute
+    text: src; for: frame; url: multipage/obsolete.html##dom-frame-src
```
This should be `type: element-attr`, and `##` should be `#`. Then you can use `<{frame/src}>` to reference it.

Hopefully we can update HTML to make this unnecessary, but that's not worth waiting on.

ulan replied:
Done.

domenic reviewed (approved):
LGTM! Found a few more minor wording nits. The only substantial thing is the CrossOriginIsolated extended attribute being missing.

syg commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+
+: <dfn for="memory attribution token">cross-origin aggregated flag</dfn>
+:: A [=boolean=] flag indicating whether the token was created for aggregating the memory usage of
+   cross-origin JavaSript realms.
```
```suggestion
   cross-origin [=JavaSript realms=].
```

ulan replied:
Done.

syg commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+:: The size of the memory that this entry describes.
+
+:  <code>breakdown . {{MemoryBreakdownEntry/attribution}}</code>
+:: An array of [=JavaScript realms=] specified by their URLs and/or container elements that use the memory.
```
Maybe "An array of the URLs and/or container elements of the [=JavaScript realms=] that use the memory". Currently it sounds like `.attribution` returns an array of Realms themselves, which aren't JS language values (they're Records, which are specification values).

ulan replied:
Done.

syg commented on index.src.html:
```diff
@@ -0,0 +1,773 @@
+<div algorithm>
+  To evaluate <dfn>memory measurement allowed predicate</dfn>
+  given a [=JavaScript realm=] |realm|:
+  1. Let |global object| be |realm|'s [=Realm/global object=].
```
How does `[=Realm/global object=]` get rendered, out of curiosity?

ulan replied:
It appears as link to https://html.spec.whatwg.org/multipage/webappapis.html#concept-realm-global

syg reviewed (approved):
I don't have much to add, it looks pretty good to my eyes, but I'm not very familiar with web specs' editorial conventions.

A clarifying question for @domenic: when web specs say "JavaScript realm", do they mean the Realm Record? Here, it's mostly passed around as an opaque type, but sometimes components are accessed like the global object. Is this defined somewhere? For example, for the global object, is this defined to mean accessing the [[GlobalObject]] slot?

domenic commented:
> A clarifying question for @domenic: when web specs say "JavaScript realm", do they mean the Realm Record?

Yep!

> Here, it's mostly passed around as an opaque type, but sometimes components are accessed like the global object. Is this defined somewhere? For example, for the global object, is this defined to mean accessing the [[GlobalObject]] slot?

Yep! https://html.spec.whatwg.org/#concept-realm-global

ulan commented:
Thanks @domenic  and @syg. I addressed the comments and am going to merge the PR. I'll follow up with "Security Considerations".
